From 501d65a127283dee425e2cf373262966d874ae9f Mon Sep 17 00:00:00 2001
From: Kaisheng Yang <kaisheng@qti.qualcomm.com>
Date: Tue, 24 Jun 2025 15:03:26 +0800
Subject: [PATCH 2/2] QCLINUX: MPAM: Snapshot of QCOM SLC MPAM Driver

Snapshot for QCOM SLC MPAM driver from msm-kernel commit
877c57af1572(" mpam: mpam_msc_slc: Removes additional copying of monitor data"). Align with SCMI SLC protocol.

Change-Id: I1fc8c86ad9c5871800bc5a9508ea5ca5556a22cc
Signed-off-by: Kaisheng Yang <kaisheng@qti.qualcomm.com>
---
 drivers/soc/qcom/mpam/mpam_msc.c     | 258 ++++----
 drivers/soc/qcom/mpam/mpam_msc_slc.c | 867 +++++++++++++++++++++------
 include/soc/qcom/mpam_msc.h          |  42 +-
 include/soc/qcom/mpam_slc.h          | 282 +++++++--
 4 files changed, 1078 insertions(+), 371 deletions(-)

diff --git a/drivers/soc/qcom/mpam/mpam_msc.c b/drivers/soc/qcom/mpam/mpam_msc.c
index 598acdf747cd..02135f5a893e 100644
--- a/drivers/soc/qcom/mpam/mpam_msc.c
+++ b/drivers/soc/qcom/mpam/mpam_msc.c
@@ -11,6 +11,7 @@
 #include <linux/of.h>
 #include <linux/of_platform.h>
 #include <linux/string.h>
+#include <linux/platform_device.h>
 #include <linux/qcom_scmi_vendor.h>
 #include <linux/scmi_protocol.h>
 #include <soc/qcom/mpam_msc.h>
@@ -54,253 +55,284 @@ static struct qcom_mpam_msc *msc_param_verification(uint32_t msc_id, struct msc_
 	return qcom_msc;
 }
 
+int msc_system_get_mpam_version(uint32_t msc_id, void *arg1)
+{
+	struct qcom_mpam_msc *qcom_msc;
+
+	if (arg1 == NULL)
+		return -EINVAL;
+
+	qcom_msc = qcom_msc_lookup(msc_id);
+	if (qcom_msc == NULL) {
+		pr_err("msc_id is not correct %d\n", msc_id);
+		return -EINVAL;
+	}
+
+	switch (qcom_msc->qcom_msc_id.qcom_msc_type) {
+	case SLC:
+		if (qcom_msc->ops->get_firmware_version)
+			return qcom_msc->ops->get_firmware_version(qcom_msc->dev, arg1);
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(msc_system_get_mpam_version);
+
 int msc_system_get_partition(uint32_t msc_id, void *arg1, void *arg2)
 {
-	int ret = -EINVAL;
-	uint8_t qcom_msc_type;
 	struct qcom_mpam_msc *qcom_msc;
-	struct msc_query *query;
 
-	query = (struct msc_query *) arg1;
-	if (query == NULL)
+	if ((arg1 == NULL) || (arg2 == NULL))
 		return -EINVAL;
 
-	qcom_msc = msc_param_verification(msc_id, query);
+	qcom_msc = msc_param_verification(msc_id, (struct msc_query *)arg1);
 	if (qcom_msc == NULL)
-		return ret;
+		return -EINVAL;
 
-	qcom_msc_type = qcom_msc->qcom_msc_id.qcom_msc_type;
-	switch (qcom_msc_type) {
+	switch (qcom_msc->qcom_msc_id.qcom_msc_type) {
 	case SLC:
 		if (qcom_msc->ops->get_cache_partition)
-			ret = qcom_msc->ops->get_cache_partition(qcom_msc->dev, arg1, arg2);
+			return qcom_msc->ops->get_cache_partition(qcom_msc->dev, arg1, arg2);
 		break;
 	default:
 		break;
 	}
 
-	return ret;
+	return 0;
 }
 EXPORT_SYMBOL_GPL(msc_system_get_partition);
 
 int msc_system_get_device_capability(uint32_t msc_id, void *arg1, void *arg2)
 {
-	int ret = -EINVAL;
-	uint8_t qcom_msc_type;
 	struct qcom_mpam_msc *qcom_msc;
-	struct msc_query *query;
 
-	query = (struct msc_query *) arg1;
-	if (query == NULL)
+	if ((arg1 == NULL) || (arg2 == NULL))
 		return -EINVAL;
 
-	qcom_msc = msc_param_verification(msc_id, query);
+	qcom_msc = msc_param_verification(msc_id, (struct msc_query *)arg1);
 	if (qcom_msc == NULL)
-		return ret;
+		return -EINVAL;
 
-	qcom_msc_type = qcom_msc->qcom_msc_id.qcom_msc_type;
-	switch (qcom_msc_type) {
+	switch (qcom_msc->qcom_msc_id.qcom_msc_type) {
 	case SLC:
 		if (qcom_msc->ops->get_cache_partition_capability)
-			ret = qcom_msc->ops->get_cache_partition_capability(qcom_msc->dev,
+			return qcom_msc->ops->get_cache_partition_capability(qcom_msc->dev,
 					arg1, arg2);
 		break;
 	default:
 		break;
 	}
 
-	return ret;
+	return 0;
 }
 EXPORT_SYMBOL_GPL(msc_system_get_device_capability);
 
 int msc_system_set_partition(uint32_t msc_id, void *arg1, void *arg2)
 {
-	int ret = -EINVAL;
-	uint8_t qcom_msc_type;
 	struct qcom_mpam_msc *qcom_msc;
-	struct msc_query *query;
 
-	query = (struct msc_query *) arg1;
-	if (query == NULL)
+	if ((arg1 == NULL) || (arg2 == NULL))
 		return -EINVAL;
 
-	qcom_msc = msc_param_verification(msc_id, query);
+	qcom_msc = msc_param_verification(msc_id, (struct msc_query *)arg1);
 	if (qcom_msc == NULL)
-		return ret;
+		return -EINVAL;
 
-	qcom_msc_type = qcom_msc->qcom_msc_id.qcom_msc_type;
-	switch (qcom_msc_type) {
+	switch (qcom_msc->qcom_msc_id.qcom_msc_type) {
 	case SLC:
 		if (qcom_msc->ops->set_cache_partition)
-			ret = qcom_msc->ops->set_cache_partition(qcom_msc->dev, arg1, arg2);
+			return qcom_msc->ops->set_cache_partition(qcom_msc->dev, arg1, arg2);
 		break;
 	default:
 		break;
 	}
 
-	return ret;
+	return 0;
 }
 EXPORT_SYMBOL_GPL(msc_system_set_partition);
 
 int msc_system_mon_config(uint32_t msc_id, void *arg1, void *arg2)
 {
-	int ret = -EINVAL;
-	uint8_t qcom_msc_type;
 	struct qcom_mpam_msc *qcom_msc;
-	struct msc_query *query;
 
-	query = (struct msc_query *) arg1;
-	if (query == NULL)
+	if ((arg1 == NULL) || (arg2 == NULL))
 		return -EINVAL;
 
-	qcom_msc = msc_param_verification(msc_id, query);
+	qcom_msc = msc_param_verification(msc_id, (struct msc_query *)arg1);
 	if (qcom_msc == NULL)
-		return ret;
+		return -EINVAL;
 
-	qcom_msc_type = qcom_msc->qcom_msc_id.qcom_msc_type;
-	switch (qcom_msc_type) {
+	switch (qcom_msc->qcom_msc_id.qcom_msc_type) {
 	case SLC:
 		if (qcom_msc->ops->mon_config)
-			ret = qcom_msc->ops->mon_config(qcom_msc->dev, arg1, arg2);
+			return qcom_msc->ops->mon_config(qcom_msc->dev, arg1, arg2);
 		break;
 	default:
 		break;
 	}
 
-	return ret;
+	return 0;
 }
 EXPORT_SYMBOL_GPL(msc_system_mon_config);
 
 int msc_system_reset_partition(uint32_t msc_id, void *arg1, void *arg2)
 {
-	int ret = -EINVAL;
-	uint8_t qcom_msc_type;
 	struct qcom_mpam_msc *qcom_msc;
-	struct msc_query *query;
 
-	query = (struct msc_query *) arg1;
-	if (query == NULL)
+	if ((arg1 == NULL) || (arg2 == NULL))
 		return -EINVAL;
 
-	qcom_msc = msc_param_verification(msc_id, query);
+	qcom_msc = msc_param_verification(msc_id, (struct msc_query *)arg1);
 	if (qcom_msc == NULL)
-		return ret;
+		return -EINVAL;
 
-	qcom_msc_type = qcom_msc->qcom_msc_id.qcom_msc_type;
-	switch (qcom_msc_type) {
+	switch (qcom_msc->qcom_msc_id.qcom_msc_type) {
 	case SLC:
 		if (qcom_msc->ops->reset_cache_partition)
-			ret = qcom_msc->ops->reset_cache_partition(qcom_msc->dev, arg1, arg2);
+			return qcom_msc->ops->reset_cache_partition(qcom_msc->dev, arg1, arg2);
 		break;
 	default:
 		break;
 	}
 
-	return ret;
+	return 0;
 }
 EXPORT_SYMBOL_GPL(msc_system_reset_partition);
 
-int msc_system_mon_alloc_info(uint32_t msc_id, void *arg1, void *arg2)
+int msc_system_mon_fe_bw_info(uint32_t msc_id, void *arg1, void *arg2)
 {
-	int part_id;
-	uint8_t qcom_msc_type;
 	struct qcom_mpam_msc *qcom_msc;
-	int ret = -EINVAL;
-	struct qcom_msc_slc_mon_val val = {0};
-	struct msc_query *query;
 	union mon_values *mon_data;
-	struct qcom_slc_mon_data_val *data;
 
+	if ((arg1 == NULL) || (arg2 == NULL))
+		return -EINVAL;
 
-	query = (struct msc_query *) arg1;
-	if (query == NULL)
+	qcom_msc = msc_param_verification(msc_id, (struct msc_query *)arg1);
+	if (qcom_msc == NULL)
 		return -EINVAL;
 
-	mon_data = (union mon_values *) arg2;
-	if (mon_data == NULL)
+	mon_data = (union mon_values *)arg2;
+	mon_data->ref.slc_mon_function = CACHE_FE_MON_CONFIG;
+	switch (qcom_msc->qcom_msc_id.qcom_msc_type) {
+	case SLC:
+		if (qcom_msc->ops->mon_stats_read)
+			return qcom_msc->ops->mon_stats_read(qcom_msc->dev, arg1, arg2);
+		break;
+	default:
+		break;
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL_GPL(msc_system_mon_fe_bw_info);
+
+int msc_system_mon_be_bw_info(uint32_t msc_id, void *arg1, void *arg2)
+{
+	struct qcom_mpam_msc *qcom_msc;
+	union mon_values *mon_data;
+
+	if ((arg1 == NULL) || (arg2 == NULL))
 		return -EINVAL;
 
-	qcom_msc = msc_param_verification(msc_id, query);
+	qcom_msc = msc_param_verification(msc_id, (struct msc_query *)arg1);
 	if (qcom_msc == NULL)
-		return ret;
+		return -EINVAL;
 
-	qcom_msc_type = qcom_msc->qcom_msc_id.qcom_msc_type;
-	switch (qcom_msc_type) {
+	mon_data = (union mon_values *)arg2;
+	mon_data->ref.slc_mon_function = CACHE_BE_MON_CONFIG;
+	switch (qcom_msc->qcom_msc_id.qcom_msc_type) {
 	case SLC:
 		if (qcom_msc->ops->mon_stats_read)
-			ret = qcom_msc->ops->mon_stats_read(qcom_msc->dev, arg1, &val);
+			return qcom_msc->ops->mon_stats_read(qcom_msc->dev, arg1, arg2);
 		break;
 	default:
 		break;
 	}
 
-	if (ret)
-		return ret;
+	return 0;
+}
+EXPORT_SYMBOL_GPL(msc_system_mon_be_bw_info);
 
-	for (part_id = 0; part_id < SLC_NUM_PARTIDS; part_id++) {
-		data = &(val.data[part_id]);
-		if ((query->client_id == data->part_info.client_id) &&
-				(query->part_id == data->part_info.part_id)) {
-			mon_data->capacity.num_cache_lines = data->num_cache_lines;
-			break;
-		}
-	}
+int msc_system_mon_alloc_info(uint32_t msc_id, void *arg1, void *arg2)
+{
+	struct qcom_mpam_msc *qcom_msc;
+	union mon_values *mon_data;
 
-	mon_data->capacity.last_capture_time = val.last_capture_time;
+	if ((arg1 == NULL) || (arg2 == NULL))
+		return -EINVAL;
+
+	qcom_msc = msc_param_verification(msc_id, (struct msc_query *)arg1);
+	if (qcom_msc == NULL)
+		return -EINVAL;
 
-	return ret;
+	mon_data = (union mon_values *)arg2;
+	mon_data->ref.slc_mon_function = CACHE_CAPACITY_CONFIG;
+	switch (qcom_msc->qcom_msc_id.qcom_msc_type) {
+	case SLC:
+		if (qcom_msc->ops->mon_stats_read)
+			return qcom_msc->ops->mon_stats_read(qcom_msc->dev, arg1, arg2);
+		break;
+	default:
+		break;
+	}
+
+	return 0;
 }
 EXPORT_SYMBOL_GPL(msc_system_mon_alloc_info);
 
 int msc_system_mon_read_miss_info(uint32_t msc_id, void *arg1, void *arg2)
 {
-	int part_id;
-	uint8_t qcom_msc_type;
 	struct qcom_mpam_msc *qcom_msc;
-	int ret = -EINVAL;
-	struct qcom_msc_slc_mon_val val = {0};
-	struct msc_query *query;
 	union mon_values *mon_data;
-	struct qcom_slc_mon_data_val *data;
 
-	query = (struct msc_query *) arg1;
-	if (query == NULL)
+	if ((arg1 == NULL) || (arg2 == NULL))
 		return -EINVAL;
 
-	mon_data = (union mon_values *) arg2;
-	if (mon_data == NULL)
-		return -EINVAL;
-
-	qcom_msc = msc_param_verification(msc_id, query);
+	qcom_msc = msc_param_verification(msc_id, (struct msc_query *)arg1);
 	if (qcom_msc == NULL)
-		return ret;
+		return -EINVAL;
 
-	qcom_msc_type = qcom_msc->qcom_msc_id.qcom_msc_type;
-	switch (qcom_msc_type) {
+	mon_data = (union mon_values *)arg2;
+	mon_data->ref.slc_mon_function = CACHE_READ_MISS_CONFIG;
+	switch (qcom_msc->qcom_msc_id.qcom_msc_type) {
 	case SLC:
 		if (qcom_msc->ops->mon_stats_read)
-			ret = qcom_msc->ops->mon_stats_read(qcom_msc->dev, arg1, &val);
+			return qcom_msc->ops->mon_stats_read(qcom_msc->dev, arg1, arg2);
 		break;
 	default:
 		break;
 	}
 
-	if (ret)
-		return ret;
+	return 0;
+}
+EXPORT_SYMBOL_GPL(msc_system_mon_read_miss_info);
 
-	for (part_id = 0; part_id < SLC_NUM_PARTIDS; part_id++) {
-		data = &(val.data[part_id]);
-		if ((query->client_id == data->part_info.client_id) &&
-				(query->part_id == data->part_info.part_id)) {
-			mon_data->misses.num_rd_misses = data->rd_misses;
-			break;
-		}
+int msc_system_mon_read_all(uint32_t msc_id, void *arg1, void *arg2)
+{
+	struct qcom_mpam_msc *qcom_msc;
+
+	if ((arg1 == NULL) || (arg2 == NULL))
+		return -EINVAL;
+
+	qcom_msc = msc_param_verification(msc_id, (struct msc_query *)arg1);
+	if (qcom_msc == NULL)
+		return -EINVAL;
+
+	switch (qcom_msc->qcom_msc_id.qcom_msc_type) {
+	case SLC:
+		if (qcom_msc->ops->mon_stats_print)
+			return qcom_msc->ops->mon_stats_print(qcom_msc->dev, arg1, arg2);
+		break;
+	default:
+		break;
 	}
 
-	mon_data->misses.last_capture_time = val.last_capture_time;
-	return ret;
+	return 0;
 }
-EXPORT_SYMBOL_GPL(msc_system_mon_read_miss_info);
+EXPORT_SYMBOL_GPL(msc_system_mon_read_all);
 
 int attach_mpam_msc(struct device *dev, struct qcom_mpam_msc *qcom_msc,
 		uint32_t msc_type)
@@ -347,7 +379,7 @@ static int mpam_msc_probe(struct platform_device *pdev)
 	return 0;
 }
 
-int mpam_msc_remove(struct platform_device *pdev)
+static int mpam_msc_remove(struct platform_device *pdev)
 {
 	return 0;
 }
diff --git a/drivers/soc/qcom/mpam/mpam_msc_slc.c b/drivers/soc/qcom/mpam/mpam_msc_slc.c
index 800093c6fd2b..30640cd4c50c 100644
--- a/drivers/soc/qcom/mpam/mpam_msc_slc.c
+++ b/drivers/soc/qcom/mpam/mpam_msc_slc.c
@@ -12,27 +12,15 @@
 #include <linux/of_platform.h>
 #include <linux/string.h>
 #include <linux/delay.h>
+#include <linux/platform_device.h>
 #include <linux/qcom_scmi_vendor.h>
 #include <linux/scmi_protocol.h>
 #include <soc/qcom/mpam_msc.h>
 #include <soc/qcom/mpam_slc.h>
+#include <linux/clk.h>
 
-#define QCOM_SLC_MPAM_SCMI_STR	0x534c434d50414d /* SLCMPAM */
-#define SCMI_GET_PARAM_LEN	32
-
-enum mpam_slc_set_param_ids {
-	PARAM_SET_CACHE_PARTITION_MSC = 1,
-	PARAM_RESET_CACHE_PARTITION_MSC = 2,
-	PARAM_SET_CONFIG_MON_MSC = 3,
-	PARAM_SET_CONFIG_SLC_MPAM_START_STOP = 4,
-};
-
-enum mpam_slc_get_param_ids {
-	PARAM_GET_CLIENT_INFO_MSC = 1,
-	PARAM_GET_CACHE_CAPABILITY_MSC = 2,
-	PARAM_GET_CACHE_PARTITION_MSC = 3,
-};
-
+#define SCMI_GET_PARAM_LEN	64
+/* Older firmware Client details */
 static uint16_t slc_client_id[] = {
 	APPS,
 	GPU,
@@ -40,6 +28,17 @@ static uint16_t slc_client_id[] = {
 	SLC_CLIENT_MAX,
 };
 
+static struct clk *llcc_perfmon_clock;
+static int perf_clk_enable;
+
+static void llcc_perf_clk_enable(int enable)
+{
+	if (enable)
+		clk_prepare_enable(llcc_perfmon_clock);
+	else
+		clk_disable_unprepare(llcc_perfmon_clock);
+}
+
 static int mpam_msc_slc_set_params(struct device *dev, void *param, int param_len,
 		uint32_t param_id)
 {
@@ -76,6 +75,15 @@ static int mpam_msc_slc_get_params(struct device *dev, void *param_in, int param
 	return ret;
 }
 
+static int qcom_mpam_msc_enabled(struct qcom_mpam_msc *qcom_msc)
+{
+	if ((qcom_msc == NULL) || (qcom_msc->msc_capability == NULL) ||
+			 (qcom_msc->mpam_available & MPAM_AVAILABLE) == 0)
+		return -EINVAL;
+
+	return 0;
+}
+
 static struct qcom_mpam_msc *slc_capability_check(struct device *dev, struct msc_query *query)
 {
 	struct qcom_mpam_msc *qcom_msc;
@@ -86,16 +94,16 @@ static struct qcom_mpam_msc *slc_capability_check(struct device *dev, struct msc
 	client_idx = query->client_id;
 	partid_idx = query->part_id;
 
-	if (client_idx >= SLC_CLIENT_MAX) {
-		dev_err(dev, "Invalid Client ID %d\n", client_idx);
-		return NULL;
-	}
-
 	qcom_msc = (struct qcom_mpam_msc *)dev_get_drvdata(dev);
-	if (qcom_msc == NULL)
+	if (qcom_mpam_msc_enabled(qcom_msc))
 		return NULL;
 
 	slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
+	if (client_idx >= slc_capability->num_clients) {
+		dev_err(dev, "Invalid Client ID %d\n", client_idx);
+		return NULL;
+	}
+
 	slc_client_cap = &(slc_capability->slc_client_cap[client_idx]);
 
 	if (slc_client_cap->enabled == false) {
@@ -111,36 +119,56 @@ static struct qcom_mpam_msc *slc_capability_check(struct device *dev, struct msc
 	return qcom_msc;
 }
 
+static int slc_mpam_version(struct device *dev, void *val)
+{
+	struct qcom_mpam_msc *qcom_msc;
+	struct qcom_slc_capability *slc_capability;
+	uint32_t *firmware_ver = val;
+
+	qcom_msc = (struct qcom_mpam_msc *)dev_get_drvdata(dev);
+	if (qcom_mpam_msc_enabled(qcom_msc))
+		return -EINVAL;
+
+	slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
+	*firmware_ver = slc_capability->firmware_ver.firmware_version;
+	return 0;
+}
+
 static int slc_set_cache_partition(struct device *dev, void *msc_partid, void *msc_partconfig)
 {
 	struct qcom_mpam_msc *qcom_msc;
 	struct slc_partid_config slc_part_config;
 	struct qcom_slc_capability *slc_capability;
 	struct slc_client_capability *slc_client_cap;
-	struct slc_partid_capability *slc_partid_cap;
+	struct slc_partid_capability *partid_cap;
 	int client_idx, partid_idx, gear_idx;
 
-	memcpy(&slc_part_config.query, msc_partid, sizeof(struct msc_query));
-	memcpy(&slc_part_config.gear_config, msc_partconfig, sizeof(struct qcom_slc_gear_val));
-
 	qcom_msc = slc_capability_check(dev, &slc_part_config.query);
 	if (qcom_msc == NULL)
 		return -EINVAL;
 
-	client_idx = slc_part_config.query.client_id;
-	partid_idx = slc_part_config.query.part_id;
 	slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
-	slc_client_cap = &(slc_capability->slc_client_cap[client_idx]);
-	slc_partid_cap = &(slc_client_cap->slc_partid_cap[partid_idx]);
-
-	for (gear_idx = 0; gear_idx < slc_partid_cap->num_gears; gear_idx++) {
-		if (slc_partid_cap->part_id_gears[gear_idx] == slc_part_config.gear_config.gear_val)
-			break;
-	}
+	memcpy(&slc_part_config.query, msc_partid, sizeof(struct msc_query));
+	slc_part_config.gear_config.gear_val =
+		((struct qcom_slc_gear_val *)msc_partconfig)->gear_val;
+
+	if (slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_0) {
+		client_idx = slc_part_config.query.client_id;
+		partid_idx = slc_part_config.query.part_id;
+		slc_client_cap = &(slc_capability->slc_client_cap[client_idx]);
+		partid_cap = (struct slc_partid_capability *)
+			&(slc_client_cap->slc_partid_cap[partid_idx]);
+
+		for (gear_idx = 0; gear_idx < partid_cap->num_gears; gear_idx++) {
+			if (partid_cap->part_id_gears[gear_idx] ==
+					slc_part_config.gear_config.gear_val)
+				break;
+		}
 
-	if (gear_idx == slc_partid_cap->num_gears) {
-		dev_err(dev, "GEAR config not valid!\n");
-		return -EINVAL;
+		if (gear_idx == partid_cap->num_gears) {
+			dev_err(dev, "GEAR config not valid!\n");
+			return -EINVAL;
+		}
 	}
 
 	return mpam_msc_slc_set_params(dev, &slc_part_config, sizeof(struct slc_partid_config),
@@ -150,19 +178,38 @@ static int slc_set_cache_partition(struct device *dev, void *msc_partid, void *m
 static int slc_reset_cache_partition(struct device *dev, void *msc_partid, void *msc_partconfig)
 {
 	struct qcom_mpam_msc *qcom_msc;
+	struct qcom_slc_capability *slc_capability;
 	struct slc_partid_config slc_part_config = {0};
 
-	memcpy(&slc_part_config.query, msc_partid, sizeof(struct msc_query));
-	memcpy(&slc_part_config.gear_config, msc_partconfig, sizeof(struct qcom_slc_gear_val));
-
 	qcom_msc = slc_capability_check(dev, &slc_part_config.query);
 	if (qcom_msc == NULL)
 		return -EINVAL;
 
+	slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
+	memcpy(&slc_part_config.query, msc_partid, sizeof(struct msc_query));
+
+	slc_part_config.gear_config.gear_val =
+		((struct qcom_slc_gear_val *)msc_partconfig)->gear_val;
+
 	return mpam_msc_slc_set_params(dev, &slc_part_config, sizeof(struct slc_partid_config),
 			PARAM_RESET_CACHE_PARTITION_MSC);
 }
 
+static int slc_firmware_version_query(struct device *dev, void *msc_partid, void *version)
+{
+	struct qcom_mpam_msc *qcom_msc;
+	struct msc_query query;
+	struct qcom_slc_firmware_version *firmware_ver;
+
+	firmware_ver = (struct qcom_slc_firmware_version *)version;
+	qcom_msc = (struct qcom_mpam_msc *)dev_get_drvdata(dev);
+	if (qcom_msc == NULL)
+		return -EINVAL;
+
+	return mpam_msc_slc_get_params(dev, &query, sizeof(struct msc_query), firmware_ver,
+			sizeof(struct qcom_slc_firmware_version), PARAM_GET_SLC_MPAM_VERSION);
+}
+
 static int slc_client_query(struct device *dev, void *msc_partid, void *msc_partconfig)
 {
 	struct qcom_mpam_msc *qcom_msc;
@@ -170,7 +217,7 @@ static int slc_client_query(struct device *dev, void *msc_partid, void *msc_part
 	struct slc_client_info *client_info;
 
 	query = (struct msc_query *)msc_partid;
-	client_info = (struct slc_client_info *) msc_partconfig;
+	client_info = (struct slc_client_info *)msc_partconfig;
 
 	qcom_msc = (struct qcom_mpam_msc *)dev_get_drvdata(dev);
 	if (qcom_msc == NULL)
@@ -184,18 +231,18 @@ static int slc_get_cache_partition(struct device *dev, void *msc_partid, void *m
 {
 	struct qcom_mpam_msc *qcom_msc;
 	struct msc_query *query;
-	struct qcom_slc_capability *qcom_slc_capability;
+	struct qcom_slc_capability *slc_capability;
 	struct slc_client_capability *slc_client_cap;
 	struct qcom_slc_gear_val *gear_config;
 
-	query = (struct msc_query *) msc_partid;
+	query = (struct msc_query *)msc_partid;
 	qcom_msc = slc_capability_check(dev, query);
 	if (qcom_msc == NULL)
 		return -EINVAL;
 
 	gear_config = (struct qcom_slc_gear_val *)msc_partconfig;
-	qcom_slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
-	slc_client_cap = &qcom_slc_capability->slc_client_cap[query->client_id];
+	slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
+	slc_client_cap = &slc_capability->slc_client_cap[query->client_id];
 	if (slc_client_cap->enabled == false)
 		return -EINVAL;
 
@@ -208,42 +255,85 @@ static int slc_get_cache_partition_capability(struct device *dev, void *msc_part
 {
 	struct qcom_mpam_msc *qcom_msc;
 	struct msc_query *query;
-	struct qcom_slc_capability *qcom_slc_capability;
-	struct slc_partid_capability *slc_partid_capability;
+	struct qcom_slc_capability *slc_capability;
+	union slc_partid_capability_def *partid_cap;
 	struct slc_client_capability *slc_client_cap;
 
-	query = (struct msc_query *) msc_partid;
+	query = (struct msc_query *)msc_partid;
 	qcom_msc = slc_capability_check(dev, query);
 	if (qcom_msc == NULL)
 		return -EINVAL;
 
-	slc_partid_capability = (struct slc_partid_capability *) msc_partconfig;
-	qcom_slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
-	slc_client_cap = &qcom_slc_capability->slc_client_cap[query->client_id];
+	partid_cap = (union slc_partid_capability_def *)msc_partconfig;
+	slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
+	slc_client_cap = &slc_capability->slc_client_cap[query->client_id];
 	if (slc_client_cap->enabled == false)
 		return -EINVAL;
 
-	return mpam_msc_slc_get_params(dev, query, sizeof(struct msc_query), slc_partid_capability,
-			sizeof(struct slc_partid_capability), PARAM_GET_CACHE_CAPABILITY_MSC);
+	return mpam_msc_slc_get_params(dev, query, sizeof(struct msc_query),
+			partid_cap, sizeof(union slc_partid_capability_def),
+			PARAM_GET_CACHE_CAPABILITY_MSC);
 }
 
-static int mon_idx_lookup(void __iomem *mem, int client_id, int part_id)
+static int mon_idx_lookup(void __iomem *mem, int client_id, int part_id,
+		struct qcom_slc_capability *slc_capability)
 {
 	int mon_idx;
-	struct qcom_slc_mon_mem *mon_mem = (struct qcom_slc_mon_mem *)mem;
+	union qcom_slc_monitor_memory *mon_mem = (union qcom_slc_monitor_memory *)mem;
 	struct slc_partid_info *part_info;
+	struct qcom_slc_mon_data *data;
+	struct qcom_slc_mon_data_v1 *data_v1;
+	struct slc_client_capability *slc_client_cap;
+	union slc_partid_capability_def *slc_partid_cap;
+	int client_idx, part_idx;
+
+	mon_idx = 0;
+	slc_client_cap = slc_capability->slc_client_cap;
+	if (slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_3)
+		data_v1 = &(mon_mem->mem_v2.data[0]);
+	else if ((slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_1) ||
+		(slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_2))
+		data = &(mon_mem->mem_v1.data[0]);
+	else
+		data = &(mon_mem->mem_v0.data[0]);
+
+	for (client_idx = 0; client_idx < slc_capability->num_clients; client_idx++) {
+		slc_partid_cap = slc_client_cap->slc_partid_cap;
+		for (part_idx = 0; part_idx < slc_client_cap->client_info.num_part_id; part_idx++) {
+			if (slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_0) {
+				part_info = &(data->part_info);
+				data++;
+			} else if (slc_capability->firmware_ver.firmware_version ==
+					SLC_MPAM_VERSION_3) {
+				if (slc_partid_cap->v1_cap.mon_support == 0) {
+					slc_partid_cap++;
+					continue;
+				}
+
+				part_info = &(data_v1->part_info);
+				data_v1++;
+			} else {
+				if (slc_partid_cap->v1_cap.mon_support == 0) {
+					slc_partid_cap++;
+					continue;
+				}
+
+				part_info = &(data->part_info);
+				data++;
+			}
+
+			if ((client_id == part_info->client_id) &&
+					(part_id == part_info->part_id))
+				return mon_idx;
+
+			mon_idx++;
+			slc_partid_cap++;
+		}
 
-	for (mon_idx = 0; mon_idx < SLC_NUM_PARTIDS; mon_idx++) {
-		part_info = &mon_mem->data[mon_idx].part_info;
-		if ((client_id ==  part_info->client_id) &&
-				(part_id ==  part_info->part_id))
-			break;
+		slc_client_cap++;
 	}
 
-	if (mon_idx == SLC_NUM_PARTIDS)
-		return -EINVAL;
-
-	return mon_idx;
+	return -EINVAL;
 }
 
 static struct qcom_mpam_msc *slc_config_request_check(struct device *dev, struct msc_query *query,
@@ -251,23 +341,16 @@ static struct qcom_mpam_msc *slc_config_request_check(struct device *dev, struct
 {
 	struct qcom_mpam_msc *qcom_msc;
 	struct qcom_slc_capability *slc_capability;
-	int client_idx, partid_idx, mon_idx;
-	void __iomem *mon_base;
+	int mon_idx;
 
 	qcom_msc = slc_capability_check(dev, query);
 	if (qcom_msc == NULL)
 		return NULL;
 
-	mon_base = (void __iomem *)qcom_msc->mon_base;
-	if (mon_base == NULL)
-		return NULL;
-
-	client_idx = query->client_id;
-	partid_idx = query->part_id;
 	slc_capability =  (struct qcom_slc_capability *)qcom_msc->msc_capability;
-
-	mon_idx = mon_idx_lookup(mon_base, client_idx, partid_idx);
-	if ((mon_idx < 0) && mon_idx >= SLC_NUM_PARTIDS)
+	mon_idx = mon_idx_lookup(qcom_msc->mon_base, query->client_id, query->part_id,
+			slc_capability);
+	if (mon_idx < 0)
 		return NULL;
 
 	switch (mon_cfg->slc_mon_function) {
@@ -309,6 +392,15 @@ static struct qcom_mpam_msc *slc_config_request_check(struct device *dev, struct
 			break;
 		}
 		break;
+
+	case CACHE_FE_MON_CONFIG:
+	case CACHE_BE_MON_CONFIG:
+		/* FE, BE Monitors enabled only for clients, not for Part Id basis */
+		if (query->part_id != 0) {
+			dev_err(dev, "Invalid Part ID %d\n", query->part_id);
+				return NULL;
+		}
+		break;
 	}
 
 	return qcom_msc;
@@ -320,23 +412,16 @@ static int update_mon_stats(struct device *dev, struct msc_query *query,
 {
 	struct qcom_mpam_msc *qcom_msc;
 	struct qcom_slc_capability *slc_capability;
-	int client_idx, partid_idx, mon_idx;
-	void __iomem *mon_base;
+	int mon_idx;
 
 	qcom_msc = slc_capability_check(dev, query);
 	if (qcom_msc == NULL)
 		return -EINVAL;
 
-	mon_base = (void __iomem *)qcom_msc->mon_base;
-	if (mon_base == NULL)
-		return -EINVAL;
-
-	client_idx = query->client_id;
-	partid_idx = query->part_id;
 	slc_capability =  (struct qcom_slc_capability *)qcom_msc->msc_capability;
-
-	mon_idx = mon_idx_lookup(mon_base, client_idx, partid_idx);
-	if ((mon_idx < 0) && mon_idx >= SLC_NUM_PARTIDS)
+	mon_idx = mon_idx_lookup(qcom_msc->mon_base, query->client_id, query->part_id,
+			slc_capability);
+	if (mon_idx < 0)
 		return -EINVAL;
 
 	switch (mon_cfg->slc_mon_function) {
@@ -384,8 +469,10 @@ static int slc_mon_config(struct device *dev, void *msc_partid, void *msc_partco
 	struct qcom_mpam_msc *qcom_msc;
 	struct msc_query *query;
 	struct slc_mon_config_val *mon_cfg_val;
+	struct slc_partid_capability_v1 partid_cap = { 0 };
 	struct slc_mon_config mon_cfg = {0};
 	int ret = -EINVAL;
+	uint32_t firmware_ver;
 
 	query = (struct msc_query *)msc_partid;
 	mon_cfg_val = (struct slc_mon_config_val *)msc_partconfig;
@@ -394,14 +481,80 @@ static int slc_mon_config(struct device *dev, void *msc_partid, void *msc_partco
 	if (qcom_msc == NULL)
 		return ret;
 
+	firmware_ver = SLC_MPAM_VERSION_0;
+	msc_system_get_mpam_version(SLC, &firmware_ver);
+	if (firmware_ver != SLC_MPAM_VERSION_0) {
+		ret = msc_system_get_device_capability(SLC, query, &partid_cap);
+		if (ret) {
+			pr_err("Failed to Config SLC Mon\n");
+			return ret;
+		}
+
+		switch (mon_cfg_val->slc_mon_function) {
+		default:
+			break;
+
+		case CACHE_CAPACITY_CONFIG:
+			if ((partid_cap.mon_support & (1 << cap_mon_support)) == 0)
+				return -EPERM;
+			break;
+
+		case CACHE_READ_MISS_CONFIG:
+			if ((partid_cap.mon_support & (1 << read_miss_mon_support)) == 0)
+				return -EPERM;
+			break;
+
+		case CACHE_FE_MON_CONFIG:
+			if (firmware_ver != SLC_MPAM_VERSION_3)
+				return -EPERM;
+
+			if ((partid_cap.mon_support & (1 << fe_mon_support)) == 0)
+				return -EPERM;
+
+			if (mon_cfg_val->enable && !perf_clk_enable)
+				llcc_perf_clk_enable(1);
+			break;
+
+		case CACHE_BE_MON_CONFIG:
+			if (firmware_ver != SLC_MPAM_VERSION_3)
+				return -EPERM;
+
+			if ((partid_cap.mon_support & (1 << be_mon_support)) == 0)
+				return -EPERM;
+
+			if (mon_cfg_val->enable && !perf_clk_enable)
+				llcc_perf_clk_enable(1);
+			break;
+
+		}
+	}
+
 	memcpy(&mon_cfg.query, msc_partid, sizeof(struct msc_query));
 	memcpy(&mon_cfg.config, msc_partconfig, sizeof(struct slc_mon_config_val));
 	ret = mpam_msc_slc_set_params(dev, &mon_cfg, sizeof(struct  slc_mon_config),
 			PARAM_SET_CONFIG_MON_MSC);
-	if (ret)
+	if (ret) {
+		if ((mon_cfg_val->slc_mon_function == CACHE_FE_MON_CONFIG) ||
+			(mon_cfg_val->slc_mon_function == CACHE_BE_MON_CONFIG)) {
+			if (mon_cfg_val->enable && !perf_clk_enable)
+				llcc_perf_clk_enable(0);
+		}
+
 		pr_err("Failed to Config SLC Mon\n");
-	else
+	} else {
+		if ((mon_cfg_val->slc_mon_function == CACHE_FE_MON_CONFIG) ||
+			(mon_cfg_val->slc_mon_function == CACHE_BE_MON_CONFIG)) {
+			if (mon_cfg_val->enable)
+				perf_clk_enable++;
+			else
+				perf_clk_enable--;
+
+			if (!mon_cfg_val->enable && !perf_clk_enable)
+				llcc_perf_clk_enable(0);
+		}
+
 		update_mon_stats(dev, query, mon_cfg_val);
+	}
 
 	return ret;
 };
@@ -409,124 +562,454 @@ static int slc_mon_config(struct device *dev, void *msc_partid, void *msc_partco
 #define MAX_SHARED_MEM_RETRY_CNT	5000
 #define MAX_SHARED_MATCH_SEQ_CNT	10
 
-static int slc_mon_shared_memread(void __iomem *mem,  struct qcom_msc_slc_mon_val *mon_buf)
+static int slc_mon_stats_read(struct device *dev, void *msc_partid, void *data)
 {
-	int part_id, retry_cnt = 0, match_seq_cnt = 0;
-	uint64_t timestamp;
-	uint32_t match_seq;
-	struct qcom_slc_mon_mem *mon_mem = (struct qcom_slc_mon_mem *)mem;
+	struct qcom_mpam_msc *qcom_msc;
+	struct qcom_slc_capability *slc_capability;
+	struct msc_query *query;
+	union qcom_slc_monitor_memory *mon_mem;
+	union mon_values *mon_data;
 	struct qcom_slc_mon_data *data_mem;
-	struct qcom_slc_mon_data_val *data_buf;
+	struct qcom_slc_mon_data_v1 *data_mem_v1;
+	int mon_idx;
+	uint32_t match_seq, retry_cnt, match_seq_cnt = 0;
+	volatile uint32_t *match_seq_ptr;
+	uint64_t *last_capture_time;
+
+	qcom_msc = (struct qcom_mpam_msc *)dev_get_drvdata(dev);
+	if (qcom_mpam_msc_enabled(qcom_msc))
+		return -EINVAL;
+
+	query = (struct msc_query *)msc_partid;
+	mon_data = (union mon_values *)data;
+	slc_capability =  (struct qcom_slc_capability *)qcom_msc->msc_capability;
+	mon_idx = mon_idx_lookup(qcom_msc->mon_base, query->client_id, query->part_id,
+			slc_capability);
+	if (mon_idx < 0)
+		return -EINVAL;
+
+	mon_mem = (union qcom_slc_monitor_memory *)qcom_msc->mon_base;
+	if (slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_3) {
+		data_mem_v1 = &(mon_mem->mem_v2.data[mon_idx]);
+		match_seq_ptr = &(mon_mem->mem_v2.match_seq);
+		last_capture_time = &(mon_mem->mem_v2.last_capture_time);
+	} else if ((slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_1) ||
+			(slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_2)) {
+		data_mem = &(mon_mem->mem_v1.data[mon_idx]);
+		match_seq_ptr = &(mon_mem->mem_v1.match_seq);
+		last_capture_time = &(mon_mem->mem_v1.last_capture_time);
+	} else {
+		data_mem = &(mon_mem->mem_v0.data[mon_idx]);
+		match_seq_ptr = &(mon_mem->mem_v0.match_seq);
+		last_capture_time = &(mon_mem->mem_v0.last_capture_time);
+	}
 
 	do {
-		while (unlikely((match_seq = mon_mem->match_seq) % 2) &&
+		retry_cnt = 0;
+		while (unlikely((match_seq = *match_seq_ptr) % 2) &&
 				(retry_cnt++ < MAX_SHARED_MEM_RETRY_CNT))
 			;
 
-		if (retry_cnt >= MAX_SHARED_MEM_RETRY_CNT)
-			return -EINVAL;
-
-		for (part_id = 0; part_id < SLC_NUM_PARTIDS; part_id++) {
-			data_buf = &(mon_buf->data[part_id]);
-			data_mem = &(mon_mem->data[part_id]);
-			data_buf->part_info.client_id = data_mem->part_info.client_id;
-			data_buf->part_info.part_id = data_mem->part_info.part_id;
-
-			if (data_mem->cap_stats.cap_enabled)
-				data_buf->num_cache_lines = data_mem->cap_stats.num_cache_lines;
+		if (retry_cnt >= MAX_SHARED_MEM_RETRY_CNT) {
+			pr_err("SLC MPAM Monitor failed. FW agent updating the same memory\n");
+			return -EIO;
+		}
 
-			if (data_mem->rd_miss_stats.miss_enabled)
-				data_buf->rd_misses = data_mem->rd_miss_stats.rd_misses;
+		/* Read as zero if monitor not enabled */
+		mon_data->ref.mon_data = 0;
+		if (slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_3) {
+
+			if ((mon_data->ref.slc_mon_function == CACHE_CAPACITY_CONFIG) &&
+					(data_mem_v1->mon_enabled & (1 << cap_mon_support)))
+				mon_data->capacity.num_cache_lines =
+					data_mem_v1->num_cache_lines;
+
+			if ((mon_data->ref.slc_mon_function == CACHE_READ_MISS_CONFIG) &&
+					(data_mem_v1->mon_enabled & (1 << read_miss_mon_support)))
+				mon_data->misses.num_rd_misses =
+					data_mem_v1->rd_misses;
+
+			if ((mon_data->ref.slc_mon_function == CACHE_FE_MON_CONFIG) &&
+					(data_mem_v1->mon_enabled & (1 << fe_mon_support)))
+				mon_data->fe_stats.slc_fe_bytes =
+					data_mem_v1->fe_rd_bytes +
+					data_mem_v1->fe_wr_bytes;
+
+			if ((mon_data->ref.slc_mon_function == CACHE_BE_MON_CONFIG) &&
+					(data_mem_v1->mon_enabled & (1 << be_mon_support)))
+				mon_data->be_stats.slc_be_bytes =
+					data_mem_v1->be_rd_bytes +
+					data_mem_v1->be_wr_bytes;
+
+		} else {
+			if ((mon_data->ref.slc_mon_function == CACHE_CAPACITY_CONFIG) &&
+					(data_mem->cap_stats.cap_enabled))
+				mon_data->capacity.num_cache_lines =
+					data_mem->cap_stats.num_cache_lines;
+
+			if ((mon_data->ref.slc_mon_function == CACHE_READ_MISS_CONFIG) &&
+					(data_mem->rd_miss_stats.miss_enabled))
+				mon_data->misses.num_rd_misses = data_mem->rd_miss_stats.rd_misses;
 		}
 
-		timestamp = mon_mem->last_capture_time;
-	} while ((match_seq != mon_mem->match_seq) &&
+		mon_data->ref.last_capture_time = *last_capture_time;
+	} while ((match_seq != *match_seq_ptr) &&
 			(match_seq_cnt++ < MAX_SHARED_MATCH_SEQ_CNT));
 
 	if (match_seq_cnt == MAX_SHARED_MATCH_SEQ_CNT)
-		return -EINVAL;
+		return -EIO;
 
-	mon_buf->last_capture_time = timestamp;
 	return 0;
 }
 
-static int slc_mon_stats_read(struct device *dev, void *msc_partid, void *mon_val)
+static ssize_t slc_mon_stats_print_data_v1(void *buf, struct qcom_slc_capability *slc_capability,
+		volatile struct qcom_slc_mon_data_v1 *data_v1,  uint64_t last_capture_time)
 {
-	struct qcom_mpam_msc *qcom_msc;
-	void __iomem *mon_base;
-	struct qcom_msc_slc_mon_val *mon_buf;
+	ssize_t len;
+	int client_idx, part_idx;
+	struct slc_client_capability *slc_client_cap;
 
-	mon_buf = (struct qcom_msc_slc_mon_val *)mon_val;
-	if (mon_buf == NULL)
-		return -EINVAL;
+	slc_client_cap = slc_capability->slc_client_cap;
+	len = scnprintf(buf, PAGE_SIZE, "timestamp=%llu\n", last_capture_time);
+	for (client_idx = 0; client_idx < slc_capability->num_clients;
+			client_idx++) {
+		for (part_idx = 0;
+				part_idx < slc_client_cap->client_info.num_part_id; part_idx++) {
+			if (slc_client_cap->slc_partid_cap[part_idx].v1_cap.mon_support == 0)
+				continue;
+
+			if (data_v1->mon_enabled) {
+				if (slc_client_cap->client_info.num_part_id == 1)
+					len += scnprintf(buf + len, PAGE_SIZE - len,
+							"%s:\n",
+							slc_client_cap->client_name);
+				else
+					len += scnprintf(buf + len, PAGE_SIZE - len,
+							"%s part %d:\n",
+							slc_client_cap->client_name,
+							data_v1->part_info.part_id);
+
+				if (data_v1->mon_enabled & (1 << cap_mon_support))
+					len += scnprintf(buf + len, PAGE_SIZE - len,
+							"cap_cnt=%d,", data_v1->num_cache_lines);
+
+				if (data_v1->mon_enabled &
+						(1 << read_miss_mon_support))
+					len += scnprintf(buf + len, PAGE_SIZE - len,
+							"miss_cnt=%llu,",
+							data_v1->rd_misses);
+
+				if (data_v1->mon_enabled & (1 << fe_mon_support))
+					len += scnprintf(buf + len, PAGE_SIZE - len,
+							"fe_bytes=%llu,",
+							data_v1->fe_rd_bytes +
+							data_v1->fe_wr_bytes);
+
+				if (data_v1->mon_enabled & (1 << be_mon_support))
+					len += scnprintf(buf + len, PAGE_SIZE - len,
+							"be_bytes=%llu,",
+							data_v1->be_rd_bytes +
+							data_v1->be_wr_bytes);
+
+				len -= 1;
+				len += scnprintf(buf + len, PAGE_SIZE - len, "\n");
+			}
+
+			data_v1++;
+		}
+
+		slc_client_cap++;
+	}
+
+	return len;
+}
+
+static ssize_t slc_mon_stats_print_data(void *buf, struct qcom_slc_capability *slc_capability,
+		volatile struct qcom_slc_mon_data *data,  uint64_t last_capture_time)
+{
+	ssize_t len;
+	int client_idx, part_idx;
+	struct slc_client_capability *slc_client_cap;
+	volatile struct slc_capacity *cap_stats;
+	volatile struct slc_read_miss_cntr *rd_miss_stats;
+
+	slc_client_cap = slc_capability->slc_client_cap;
+	len = scnprintf(buf, PAGE_SIZE, "timestamp=%llu\n", last_capture_time);
+	for (client_idx = 0; client_idx < slc_capability->num_clients;
+			client_idx++) {
+		for (part_idx = 0;
+				part_idx < slc_client_cap->client_info.num_part_id;
+				part_idx++) {
+			cap_stats = &data->cap_stats;
+			rd_miss_stats = &data->rd_miss_stats;
+
+			if ((slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_1) ||
+					(slc_capability->firmware_ver.firmware_version ==
+					 SLC_MPAM_VERSION_2))
+				if (slc_client_cap->slc_partid_cap[part_idx].v1_cap.mon_support
+						== 0)
+					continue;
+
+			if (cap_stats->cap_enabled ||
+					rd_miss_stats->miss_enabled) {
+				if (slc_client_cap->client_info.num_part_id == 1)
+					len += scnprintf(buf + len, PAGE_SIZE - len,
+							"%s:\n",
+							slc_client_cap->client_name);
+				else
+					len += scnprintf(buf + len, PAGE_SIZE - len,
+							"%s part %d:\n",
+							slc_client_cap->client_name,
+							data->part_info.part_id);
+
+				if (cap_stats->cap_enabled)
+					len += scnprintf(buf + len, PAGE_SIZE - len,
+							"cap_cnt=%d,",
+							cap_stats->num_cache_lines);
+
+				if (rd_miss_stats->miss_enabled)
+					len += scnprintf(buf + len, PAGE_SIZE - len,
+							"miss_cnt=%llu,",
+							rd_miss_stats->rd_misses);
+
+				len -= 1;
+				len += scnprintf(buf + len, PAGE_SIZE - len, "\n");
+			}
+
+			data++;
+		}
+
+		slc_client_cap++;
+	}
+
+	return len;
+}
+
+static int slc_mon_stats_print(struct device *dev, void *msc_partid, void *buf)
+{
+	ssize_t len;
+	uint32_t match_seq, retry_cnt, match_seq_cnt = 0;
+	struct qcom_mpam_msc *qcom_msc;
+	struct qcom_slc_capability *slc_capability;
+	union qcom_slc_monitor_memory *mon_mem;
+	volatile struct qcom_slc_mon_data *data;
+	volatile struct qcom_slc_mon_data_v1 *data_v1;
+	volatile uint32_t *match_seq_ptr;
+	volatile uint64_t *last_capture_time;
 
 	qcom_msc = (struct qcom_mpam_msc *)dev_get_drvdata(dev);
-	if (qcom_msc == NULL)
+	if (qcom_mpam_msc_enabled(qcom_msc))
 		return -EINVAL;
 
-	mon_base = (void __iomem *)qcom_msc->mon_base;
+	slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
+	mon_mem = (union qcom_slc_monitor_memory *)qcom_msc->mon_base;
+
+	if (slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_3) {
+		match_seq_ptr = &(mon_mem->mem_v2.match_seq);
+		last_capture_time = &(mon_mem->mem_v2.last_capture_time);
+	} else if ((slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_1) ||
+			(slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_2)) {
+		match_seq_ptr = &(mon_mem->mem_v1.match_seq);
+		last_capture_time = &(mon_mem->mem_v1.last_capture_time);
+	} else {
+		match_seq_ptr = &(mon_mem->mem_v0.match_seq);
+		last_capture_time = &(mon_mem->mem_v0.last_capture_time);
+	}
+
+	do {
+		retry_cnt = 0;
+		if (slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_3)
+			data_v1 = &(mon_mem->mem_v2.data[0]);
+		else if ((slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_1) ||
+			(slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_2))
+			data = &(mon_mem->mem_v1.data[0]);
+		else
+			data = &(mon_mem->mem_v0.data[0]);
+
+		while (unlikely((match_seq = *match_seq_ptr) % 2)
+				&& (retry_cnt++ < MAX_SHARED_MEM_RETRY_CNT))
+			;
 
-	slc_mon_shared_memread(mon_base, mon_buf);
+		if (unlikely(retry_cnt >= MAX_SHARED_MEM_RETRY_CNT)) {
+			pr_err("SLC MPAM Monitor failed. FW agent updating the same memory\n");
+			return -EIO;
+		}
 
-	return 0;
-};
+		if (slc_capability->firmware_ver.firmware_version == SLC_MPAM_VERSION_3)
+			len = slc_mon_stats_print_data_v1(buf, slc_capability, data_v1,
+					*last_capture_time);
+		else
+			len = slc_mon_stats_print_data(buf, slc_capability, data,
+					*last_capture_time);
+
+	} while ((match_seq != *match_seq_ptr) && (match_seq_cnt++ < MAX_SHARED_MATCH_SEQ_CNT));
+
+	if (match_seq_cnt == MAX_SHARED_MATCH_SEQ_CNT)
+		return -EIO;
+
+	return len;
+}
 
 static struct mpam_msc_ops slc_msc_ops = {
+	.get_firmware_version = slc_mpam_version,
 	.set_cache_partition = slc_set_cache_partition,
 	.get_cache_partition = slc_get_cache_partition,
 	.get_cache_partition_capability = slc_get_cache_partition_capability,
 	.reset_cache_partition = slc_reset_cache_partition,
 	.mon_config = slc_mon_config,
 	.mon_stats_read = slc_mon_stats_read,
+	.mon_stats_print = slc_mon_stats_print,
 };
 
 static int slc_client_info_read(struct device *dev, struct device_node *node)
 {
-	int client_idx, partid_idx, ret = -EINVAL;
+	int client_idx, partid_idx, num_part_ids, ret = -EINVAL;
 	struct msc_query query;
 	struct qcom_mpam_msc *qcom_msc;
-	struct slc_client_capability *slc_client_cap;
-	struct qcom_slc_capability *qcom_slc_capability;
+	struct slc_client_capability *client_cap;
+	struct qcom_slc_capability *slc_capability;
+	struct qcom_slc_mon_mem_v1 *mon_mem;
+	struct slc_sct_client_info *client_info;
+	struct slc_client_details *client_details;
 
 	qcom_msc = (struct qcom_mpam_msc *)dev_get_drvdata(dev);
 	if (qcom_msc->qcom_msc_id.qcom_msc_type != SLC)
 		return -EINVAL;
 
-	qcom_slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
+	slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
 	query.qcom_msc_id.qcom_msc_type = qcom_msc->qcom_msc_id.qcom_msc_type;
 	query.qcom_msc_id.qcom_msc_class = qcom_msc->qcom_msc_id.qcom_msc_class;
 	query.qcom_msc_id.idx = qcom_msc->qcom_msc_id.idx;
-	for (client_idx = 0; client_idx < qcom_slc_capability->num_clients; client_idx++) {
-		slc_client_cap = &(qcom_slc_capability->slc_client_cap[client_idx]);
-		query.client_id = slc_client_id[client_idx];
-		slc_client_cap->client_info.client_id = slc_client_id[client_idx];
-		ret = of_property_read_string_index(node, "qcom,slc-clients", client_idx,
-							&slc_client_cap->client_name);
-		slc_client_cap->enabled = false;
-		if (slc_client_query(dev, &query, &(slc_client_cap->client_info)))
-			continue;
-
-		if ((slc_client_cap->client_info.num_part_id == 0) ||
-				(slc_client_cap->client_info.num_part_id == SLC_INVALID_PARTID))
-			continue;
-
-		slc_client_cap->enabled = true;
-		slc_client_cap->slc_partid_cap = devm_kcalloc(dev,
-				slc_client_cap->client_info.num_part_id,
-				sizeof(struct slc_partid_capability), GFP_KERNEL);
-		if (slc_client_cap->slc_partid_cap == NULL) {
-			ret = -ENOMEM;
-			break;
+	if (slc_capability->firmware_ver.firmware_version != SLC_MPAM_VERSION_0) {
+		mon_mem = (struct qcom_slc_mon_mem_v1 *)qcom_msc->mon_base;
+		client_info = (struct slc_sct_client_info *)&(mon_mem->data);
+		slc_capability->num_clients = client_info->num_clients;
+		slc_capability->slc_mon_list.capacity_config_available =
+			client_info->slc_mon_info.num_cap_monitor;
+		slc_capability->slc_mon_list.read_miss_config_available =
+			client_info->slc_mon_info.num_miss_monitor;
+		if ((slc_capability->num_clients == 0) ||
+				(client_info->slc_mon_info.num_cap_monitor == 0) ||
+				(client_info->slc_mon_info.num_miss_monitor == 0)) {
+			pr_err("SLC Client info population Failed!\n");
+			return -EINVAL;
 		}
 
-		for (partid_idx = 0; partid_idx < slc_client_cap->client_info.num_part_id;
-				partid_idx++) {
-			query.part_id = partid_idx;
-			ret = msc_system_get_device_capability(qcom_msc->msc_id, &query,
-					&(slc_client_cap->slc_partid_cap[partid_idx]));
-			if (ret)
+		pr_info("SLC Monitor size %lld\n", mon_mem->slc_mpam_monitor_size);
+		pr_info("Number of SLC MPAM clients %d\n", slc_capability->num_clients);
+		slc_capability->slc_client_cap = devm_kcalloc(dev, slc_capability->num_clients,
+				sizeof(struct slc_client_capability), GFP_KERNEL);
+		if (slc_capability->slc_client_cap == NULL)
+			return -ENOMEM;
+
+		qcom_msc->mpam_available = MPAM_AVAILABLE;
+		client_details = &(client_info->client);
+		slc_capability->num_partids = 0;
+		for (client_idx = 0; client_idx < slc_capability->num_clients; client_idx++,
+				client_details++) {
+			client_cap = &(slc_capability->slc_client_cap[client_idx]);
+			query.client_id = client_idx;
+			query.part_id = 0;
+			client_cap->enabled = false;
+			if (slc_client_query(dev, &query, &(client_cap->client_info))) {
+				pr_err("SLC MAPM ClientID %d Query Failed!\n", query.client_id);
+				continue;
+			}
+
+			num_part_ids = client_cap->client_info.num_part_id;
+			if ((num_part_ids == 0) || (num_part_ids == SLC_INVALID_PARTID)) {
+				pr_err("SLC MAPM ClientID %d, has no valid PART ID!\n",
+						query.client_id);
 				continue;
+			}
+
+			client_cap->client_info.client_id = query.client_id;
+			client_cap->client_name = devm_kzalloc(dev, CLIENT_NAME_LEN, GFP_KERNEL);
+
+			client_cap->slc_partid_cap = devm_kcalloc(dev, num_part_ids,
+					sizeof(union slc_partid_capability_def), GFP_KERNEL);
+			if ((client_cap->client_name == NULL) ||
+					(client_cap->slc_partid_cap == NULL))
+				return -ENOMEM;
+
+			strscpy(client_cap->client_name, client_details->client_name,
+					CLIENT_NAME_LEN);
+			client_cap->enabled = true;
+			slc_capability->num_partids += num_part_ids;
+			for (partid_idx = 0; partid_idx < num_part_ids; partid_idx++) {
+				query.part_id = partid_idx;
+				ret = msc_system_get_device_capability(qcom_msc->msc_id, &query,
+						&(client_cap->slc_partid_cap[partid_idx]));
+				if (ret) {
+					client_cap->enabled = false;
+					pr_err("SLC MAPM ClientID %d, Part ID %d, failed!\n",
+							client_cap->client_info.client_id,
+							query.part_id);
+					continue;
+				}
+
+				pr_info("Client Name:%-12s Idx:%d, PartID:%d Monitor support %x\n",
+					client_details->client_name,
+					client_cap->client_info.client_id,
+					query.part_id,
+					client_cap->slc_partid_cap[partid_idx].v1_cap.mon_support);
+			}
+
 		}
+	} else {
+		slc_capability->num_clients = of_property_count_strings(node, "qcom,slc-clients");
+		pr_info("Number of SLC MPAM clients %d\n", slc_capability->num_clients);
+		slc_capability->slc_client_cap = devm_kcalloc(dev,
+				slc_capability->num_clients, sizeof(struct slc_client_capability),
+				GFP_KERNEL);
+		if (slc_capability->slc_client_cap == NULL)
+			return -ENOMEM;
+
+		if (of_property_read_u32(node, "qcom,num-read-miss-cfg",
+					&slc_capability->slc_mon_list.read_miss_config_available))
+			return -EINVAL;
+
+		if (of_property_read_u32(node, "qcom,num-cap-cfg",
+					&slc_capability->slc_mon_list.capacity_config_available))
+			return -EINVAL;
+
+		qcom_msc->mpam_available = MPAM_AVAILABLE;
+		for (client_idx = 0; client_idx < slc_capability->num_clients; client_idx++) {
+			client_cap = &(slc_capability->slc_client_cap[client_idx]);
+			client_cap->client_info.client_id = slc_client_id[client_idx];
+
+			client_cap->enabled = false;
+			query.client_id = slc_client_id[client_idx];
+			query.part_id = 0;
+			if (slc_client_query(dev, &query, &(client_cap->client_info)))
+				continue;
+
+			num_part_ids = client_cap->client_info.num_part_id;
+			if ((num_part_ids == 0) || (num_part_ids == SLC_INVALID_PARTID))
+				continue;
+
+			client_cap->enabled = true;
+			client_cap->slc_partid_cap = devm_kcalloc(dev, num_part_ids,
+					sizeof(union slc_partid_capability_def), GFP_KERNEL);
+			if (client_cap->slc_partid_cap == NULL) {
+				ret = -ENOMEM;
+				break;
+			}
+
+			for (partid_idx = 0; partid_idx < num_part_ids; partid_idx++) {
+				query.part_id = partid_idx;
+				ret = msc_system_get_device_capability(qcom_msc->msc_id, &query,
+						&(client_cap->slc_partid_cap[partid_idx]));
+				if (ret)
+					continue;
+
+				pr_info("SLC Client Idx:%d, PartID Idx:%d enabled\n",
+						client_cap->client_info.client_id,
+						query.part_id);
+			}
+		}
+
+		slc_capability->num_partids = SLC_NUM_PARTIDS;
 	}
 
 	return ret;
@@ -543,7 +1026,7 @@ static int slc_mpam_enable_disable(struct device *dev, void *msc_partid, void *m
 			PARAM_SET_CONFIG_SLC_MPAM_START_STOP);
 }
 
-static int slc_mpam_start_stop(struct device *dev, bool start)
+static int slc_mpam_start_stop(struct device *dev, uint32_t start)
 {
 	struct qcom_mpam_msc *qcom_msc;
 	struct msc_query query = {0};
@@ -565,11 +1048,11 @@ static int slc_mpam_start_stop(struct device *dev, bool start)
 static int mpam_msc_slc_probe(struct platform_device *pdev)
 {
 	int ret = -EINVAL;
-	uint32_t val;
+	uint32_t val, firmware_ver;
 	struct device_node *node;
 	struct qcom_mpam_msc *qcom_msc;
 	struct qcom_slc_capability *qcom_slc_capability;
-	struct qcom_slc_mon_mem *mon_mem;
+	union qcom_slc_monitor_memory *mon_mem;
 	struct resource *res;
 
 	qcom_msc = devm_kzalloc(&pdev->dev, sizeof(struct qcom_mpam_msc), GFP_KERNEL);
@@ -587,6 +1070,12 @@ static int mpam_msc_slc_probe(struct platform_device *pdev)
 		return dev_err_probe(&pdev->dev, PTR_ERR(qcom_msc->scmi_ops),
 				"Error getting vendor protocol ops\n");
 
+	llcc_perfmon_clock = devm_clk_get(&pdev->dev, "qdss_clk");
+	if (IS_ERR_OR_NULL(llcc_perfmon_clock)) {
+		pr_warn("failed to get qdss clock node\n");
+		llcc_perfmon_clock = NULL;
+	}
+
 	qcom_msc->qcom_msc_id.qcom_msc_type = SLC;
 	qcom_msc->qcom_msc_id.qcom_msc_class = CACHE_TYPE;
 	qcom_msc->ops = &slc_msc_ops;
@@ -597,7 +1086,7 @@ static int mpam_msc_slc_probe(struct platform_device *pdev)
 	if (ret)
 		return dev_err_probe(&pdev->dev, ret, "qcom,dev-index Node failed!\n");
 
-	if (val > 255)
+	if (val > U8_MAX)
 		return dev_err_probe(&pdev->dev, -EINVAL, "dev-index wrong value failed!\n");
 
 	qcom_msc->qcom_msc_id.idx = val;
@@ -610,8 +1099,18 @@ static int mpam_msc_slc_probe(struct platform_device *pdev)
 		return dev_err_probe(&pdev->dev, ret, "qcom,msc-id Node failed!\n");
 
 	platform_set_drvdata(pdev, qcom_msc);
-	if (slc_mpam_start_stop(&pdev->dev, true))
-		return dev_err_probe(&pdev->dev, -EINVAL, "MAPM SCMI failure!\n");
+	/* SCMI firmware query firmware version details */
+	if (slc_firmware_version_query(&pdev->dev, NULL, &firmware_ver))
+		firmware_ver = SLC_MPAM_VERSION_0;
+
+	/* Depending on firmware version SLC Monitor may memory program here */
+	val = mpam_slc_mpam_init_v0;
+	if (firmware_ver)
+		val = mpam_slc_client_info_v1;
+
+	ret = slc_mpam_start_stop(&pdev->dev, val);
+	if (ret)
+		return dev_err_probe(&pdev->dev, ret, "MAPM SCMI failure!\n");
 
 	res = platform_get_resource_byname(pdev, IORESOURCE_MEM, "mon-base");
 	qcom_msc->mon_base = devm_ioremap_resource(&pdev->dev, res);
@@ -620,8 +1119,8 @@ static int mpam_msc_slc_probe(struct platform_device *pdev)
 		return dev_err_probe(&pdev->dev, PTR_ERR(qcom_msc->mon_base),
 				"ioremap slc mpam_mon_base failed!\n");
 
-	mon_mem = (struct qcom_slc_mon_mem *) qcom_msc->mon_base;
-	if (mon_mem->msc_id != qcom_msc->msc_id)
+	mon_mem = (union qcom_slc_monitor_memory *) qcom_msc->mon_base;
+	if (mon_mem->mem_v1.msc_id != qcom_msc->msc_id)
 		return dev_err_probe(&pdev->dev, -EINVAL, "IMEM Init failure!\n");
 
 	ret = attach_mpam_msc(&pdev->dev, qcom_msc, SLC);
@@ -636,42 +1135,30 @@ static int mpam_msc_slc_probe(struct platform_device *pdev)
 	}
 
 	qcom_slc_capability = (struct qcom_slc_capability *)qcom_msc->msc_capability;
-	qcom_slc_capability->num_clients = of_property_count_strings(node, "qcom,slc-clients");
-	qcom_slc_capability->slc_client_cap = devm_kcalloc(&pdev->dev,
-			qcom_slc_capability->num_clients, sizeof(struct  slc_client_capability),
-			GFP_KERNEL);
-	if (qcom_slc_capability->slc_client_cap == NULL) {
-		ret = -ENOMEM;
+	qcom_slc_capability->firmware_ver.firmware_version = firmware_ver;
+	ret = slc_client_info_read(&pdev->dev, node);
+	if (ret) {
+		dev_err(&pdev->dev, "Failed to detect SLC device\n");
 		goto err_detach;
 	}
 
-	ret = of_property_read_u32(node, "qcom,num-read-miss-cfg", &val);
-	if (ret)
-		goto err_detach;
-
-	qcom_slc_capability->slc_mon_list.read_miss_config_available = val;
-	ret = of_property_read_u32(node, "qcom,num-cap-cfg", &val);
-	if (ret)
-		goto err_detach;
-
-	qcom_slc_capability->slc_mon_list.capacity_config_available = val;
+	if (qcom_slc_capability->firmware_ver.firmware_version != SLC_MPAM_VERSION_0) {
+		/* SLC Monitor memory programmed here on updated firmware */
+		ret = slc_mpam_start_stop(&pdev->dev, mpam_slc_mon_init_v1);
+		if (ret) {
+			dev_err(&pdev->dev, "MON Memory init Failure!\n");
+			goto err_detach;
+		}
+	}
 
 	qcom_slc_capability->slc_mon_configured.read_miss_configured = 0;
 	qcom_slc_capability->slc_mon_configured.capacity_configured = 0;
-
-	qcom_msc->mpam_available = MPAM_AVAILABLE;
-	if (slc_client_info_read(&pdev->dev, node)) {
-		dev_err(&pdev->dev, "Failed to detect SLC device\n");
-		goto err_detach;
-	}
-
 	qcom_msc->mpam_available = MPAM_MONITRS_AVAILABLE;
 	return 0;
 
 err_detach:
 	detach_mpam_msc(&pdev->dev, qcom_msc, SLC);
-	return dev_err_probe(&pdev->dev, PTR_ERR(qcom_msc->mon_base),
-			"MPAM SLC driver probe failed!\n");
+	return dev_err_probe(&pdev->dev, ret, "MPAM SLC driver probe failed!\n");
 }
 
 static int mpam_msc_slc_remove(struct platform_device *pdev)
@@ -681,7 +1168,7 @@ static int mpam_msc_slc_remove(struct platform_device *pdev)
 	qcom_msc = (struct qcom_mpam_msc *)platform_get_drvdata(pdev);
 	qcom_msc->mpam_available = MPAM_AVAILABLE;
 	if (qcom_msc != NULL) {
-		if (slc_mpam_start_stop(&pdev->dev, false))
+		if (slc_mpam_start_stop(&pdev->dev, mpam_slc_reset))
 			dev_err(&pdev->dev, "Failed to stop SLC Monitor thread\n");
 
 		detach_mpam_msc(&pdev->dev, qcom_msc, SLC);
diff --git a/include/soc/qcom/mpam_msc.h b/include/soc/qcom/mpam_msc.h
index 356a6c817985..8b040d327bf8 100644
--- a/include/soc/qcom/mpam_msc.h
+++ b/include/soc/qcom/mpam_msc.h
@@ -12,7 +12,7 @@
 enum msc_qcom_id {
 	L2,
 	L3,
-	SLC = 2,
+	SLC = 4,
 	QCOM_MSC_MAX,
 };
 
@@ -62,16 +62,22 @@ struct mpam_msc_ops {
 	int (*reset_cache_partition)(struct device *dev, void *msc_partid, void *msc_partconfig);
 	int (*mon_config)(struct device *dev, void *msc_partid, void *msc_partconfig);
 	int (*mon_stats_read)(struct device *dev, void *msc_partid, void *mon_val);
+	int (*mon_stats_print)(struct device *dev, void *msc_partid, void *mon_val);
+	int (*get_firmware_version)(struct device *dev, void *val);
 };
 
 #if IS_ENABLED(CONFIG_QTI_MPAM_MSC)
 struct qcom_mpam_msc *qcom_msc_lookup(uint32_t msc_id);
+int msc_system_get_mpam_version(uint32_t msc_id, void *arg1);
 int msc_system_get_device_capability(uint32_t msc_id, void *arg1, void *arg2);
 int msc_system_get_partition(uint32_t msc_id, void *arg1, void *arg2);
 int msc_system_set_partition(uint32_t msc_id, void *arg1, void *arg2);
 int msc_system_reset_partition(uint32_t msc_id, void *arg1, void *arg2);
 int msc_system_mon_read_miss_info(uint32_t msc_id, void *arg1, void *arg2);
 int msc_system_mon_alloc_info(uint32_t msc_id, void *arg1, void *arg2);
+int msc_system_mon_fe_bw_info(uint32_t msc_id, void *arg1, void *arg2);
+int msc_system_mon_be_bw_info(uint32_t msc_id, void *arg1, void *arg2);
+int msc_system_mon_read_all(uint32_t msc_id, void *arg1, void *arg2);
 int msc_system_mon_config(uint32_t msc_id, void *arg1, void *arg2);
 int attach_mpam_msc(struct device *dev, struct qcom_mpam_msc *qcom_msc, uint32_t msc_type);
 void detach_mpam_msc(struct device *dev, struct qcom_mpam_msc *qcom_msc, uint32_t msc_type);
@@ -82,43 +88,63 @@ static inline struct qcom_mpam_msc *qcom_msc_lookup(uint32_t msc_id)
 	return NULL;
 }
 
+static int msc_system_get_mpam_version(uint32_t msc_id, void *arg1)
+{
+	return -EINVAL;
+}
+
 static inline int msc_system_get_device_capability(uint32_t msc_id, void *arg1, void *arg2)
 {
 	return -EINVAL;
 }
 
-static inline int msc_system_get_partition(uint32_t msc_id, void *arg1, void *arg2);
+static inline int msc_system_get_partition(uint32_t msc_id, void *arg1, void *arg2)
 {
 	return -EINVAL;
 }
 
-static inline int msc_system_set_partition(uint32_t msc_id, void *arg1, void *arg2);
+static inline int msc_system_set_partition(uint32_t msc_id, void *arg1, void *arg2)
 {
 	return -EINVAL;
 }
 
-static inline int msc_system_reset_partition(uint32_t msc_id, void *arg1, void *arg2);
+static inline int msc_system_reset_partition(uint32_t msc_id, void *arg1, void *arg2)
 {
 	return -EINVAL;
 }
 
-static inline int msc_system_mon_read_miss_info(uint32_t msc_id, void *arg1, void *arg2);
+static inline int msc_system_mon_read_miss_info(uint32_t msc_id, void *arg1, void *arg2)
 {
 	return -EINVAL;
 }
 
-static inline int msc_system_mon_alloc_info(uint32_t msc_id, void *arg1, void *arg2);
+static inline int msc_system_mon_alloc_info(uint32_t msc_id, void *arg1, void *arg2)
 {
 	return -EINVAL;
 }
 
-static inline int msc_system_mon_config(uint32_t msc_id, void *arg1, void *arg2);
+static inline int msc_system_mon_fe_bw_info(uint32_t msc_id, void *arg1, void *arg2)
+{
+	return -EINVAL;
+}
+
+static inline int msc_system_mon_be_bw_info(uint32_t msc_id, void *arg1, void *arg2)
+{
+	return -EINVAL;
+}
+
+static inline int msc_system_mon_read_all(uint32_t msc_id, void *arg1, void *arg2)
+{
+	return -EINVAL;
+}
+
+static inline int msc_system_mon_config(uint32_t msc_id, void *arg1, void *arg2)
 {
 	return -EINVAL;
 }
 
 static inline int attach_mpam_msc(struct device *dev, struct qcom_mpam_msc *qcom_msc,
-		uint32_t msc_type);
+		uint32_t msc_type)
 {
 	return -EINVAL;
 }
diff --git a/include/soc/qcom/mpam_slc.h b/include/soc/qcom/mpam_slc.h
index e78d4e38a3d4..b122670e336a 100644
--- a/include/soc/qcom/mpam_slc.h
+++ b/include/soc/qcom/mpam_slc.h
@@ -6,65 +6,114 @@
 #ifndef _QCOM_MPAM_SLC_H
 #define _QCOM_MPAM_SLC_H
 
-#define MAX_NUM_GEARS		3
-#define MAX_PART_ID		10
+enum slc_clients_id {
+	APPS,
+	GPU,
+	NSP,
+	SLC_CLIENT_MAX,
+};
+
+enum slc_gears {
+	SLC_GEAR_HIGH,
+	SLC_GEAR_LOW,
+	SLC_GEAR_BYPASS,
+	MAX_NUM_GEARS,
+};
+
+static char gear_index[][25] = {
+	"SLC_GEAR_HIGH",
+	"SLC_GEAR_LOW",
+	"SLC_GEAR_BYPASS",
+	"",
+};
+
+#define CLIENT_NAME_LEN         16
+
+#define QCOM_SLC_MPAM_SCMI_STR	0x534c434d50414d /* SLCMPAM */
 #define SLC_INVALID_PARTID      ((1 << 16) - 1)
-#define SLC_NUM_PARTIDS		5
 
-/* slc Monitor capability */
-struct slc_mon_capability {
-	uint32_t read_miss_config_available;
-	uint32_t capacity_config_available;
+#define SLC_MPAM_VERSION_0	0x00000000	/* Base firmware */
+#define SLC_MPAM_VERSION_1	0x00000001	/* SLC MPAM More gear support */
+#define SLC_MPAM_VERSION_2	0x00000002	/* SLC MPAM More gear support */
+#define SLC_MPAM_VERSION_3	0x00000003	/* SLC MPAM iBW Monitor */
+
+enum mpam_slc_get_param_ids {
+	PARAM_GET_CLIENT_INFO_MSC = 1,
+	PARAM_GET_CACHE_CAPABILITY_MSC = 2,
+	PARAM_GET_CACHE_PARTITION_MSC = 3,
+	PARAM_GET_SLC_MPAM_VERSION = 4,
 };
 
-struct slc_mon_configured {
-	uint32_t read_miss_configured;
-	uint32_t capacity_configured;
+enum mpam_slc_set_param_ids {
+	PARAM_SET_CACHE_PARTITION_MSC = 1,
+	PARAM_RESET_CACHE_PARTITION_MSC = 2,
+	PARAM_SET_CONFIG_MON_MSC = 3,
+	PARAM_SET_CONFIG_SLC_MPAM_START_STOP = 4,
 };
 
-/* slc device capability */
+/* GET_PARAM */
+/* PARAM_GET_CLIENT_INFO_MSC  */
+struct slc_client_info {
+	uint16_t client_id;
+	uint16_t num_part_id;
+} __packed;
 
+/* PARAM_GET_CACHE_CAPABILITY_MSC */
 struct slc_partid_capability {
 	uint8_t part_id;
 	uint8_t num_gears;
 	uint8_t part_id_gears[MAX_NUM_GEARS];
 } __packed;
 
-struct slc_client_info {
-	uint16_t client_id;
-	uint16_t num_part_id;
+struct slc_partid_capacity_config {
+	uint32_t gear_flds_bitmap;
+	uint16_t dflt_bitmap;
+	uint32_t slc_bitfield_capacity;
 } __packed;
 
-struct slc_client_capability {
-	struct slc_client_info client_info;
-	struct slc_partid_capability *slc_partid_cap;
-	uint8_t enabled;
-	const char *client_name;
-} __packed;
+enum slc_mintor_support {
+	cap_mon_support,
+	read_miss_mon_support,
+	fe_mon_support,
+	be_mon_support,
+};
 
-struct qcom_slc_capability {
-	uint32_t num_clients;
-	struct slc_client_capability *slc_client_cap;
-	struct slc_mon_capability slc_mon_list;
-	struct slc_mon_configured slc_mon_configured;
+struct slc_partid_capability_v1 {
+	uint8_t part_id;
+	uint8_t num_gears;
+	struct slc_partid_capacity_config cap_cfg;
+	uint32_t mon_support;
 } __packed;
 
-/* slc slice configuration */
+union slc_partid_capability_def {
+	struct slc_partid_capability v0_cap;
+	struct slc_partid_capability_v1 v1_cap;
+} __packed;
 
+/* PARAM_GET_CACHE_PARTITION_MSC */
 struct qcom_slc_gear_val {
 	uint32_t gear_val;
 } __packed;
 
+/* PARAM_GET_SLC_MPAM_VERSION */
+struct qcom_slc_firmware_version {
+	uint32_t firmware_version;
+} __packed;
+
+/* SET_PARAM */
+/* PARAM_SET_CACHE_PARTITION_MSC  */
+/* PARAM_RESET_CACHE_PARTITION_MSC */
 struct slc_partid_config {
 	struct msc_query query;
 	struct qcom_slc_gear_val gear_config;
 } __packed;
 
-/* slc mon configuration */
-
+/* PARAM_SET_CONFIG_MON_MSC */
 enum slc_mon_function {
 	CACHE_CAPACITY_CONFIG,
 	CACHE_READ_MISS_CONFIG,
+	CACHE_FE_MON_CONFIG,
+	CACHE_BE_MON_CONFIG,
 };
 
 struct slc_mon_config_val {
@@ -78,6 +127,13 @@ struct slc_mon_config {
 } __packed;
 
 /* PARAM_SET_CONFIG_SLC_MPAM_START_STOP */
+enum mpam_enable_val {
+	mpam_slc_reset = 0,
+	mpam_slc_mpam_init_v0 = 1,
+	mpam_slc_client_info_v1 = 2,
+	mpam_slc_mon_init_v1 = 3,
+};
+
 struct mpam_enable {
 	uint32_t value;
 } __packed;
@@ -87,7 +143,7 @@ struct qcom_slc_mpam_enable_cfg {
 	struct mpam_enable enable;
 } __packed;
 
-/* slc monitor shared memory */
+/* shared memory SLC monitor */
 struct slc_capacity {
 	uint32_t num_cache_lines;
 	uint32_t cap_enabled;
@@ -99,18 +155,79 @@ struct slc_read_miss_cntr {
 	uint32_t miss_enabled;
 } __packed;
 
+struct slc_fe_bw_info {
+	uint64_t slc_fe_bytes;
+	uint32_t cap_enabled;
+} __packed;
+
+struct slc_be_bw_info {
+	uint64_t slc_be_bytes;
+	uint32_t cap_enabled;
+} __packed;
+
 struct slc_partid_info {
 	uint32_t client_id;
 	uint32_t part_id;
 } __packed;
 
+struct slc_client_details {
+	uint16_t client_id;
+	char client_name[CLIENT_NAME_LEN];
+} __packed;
+
+struct slc_mon_details {
+	uint32_t num_cap_monitor;
+	uint32_t num_miss_monitor;
+	uint32_t num_slc_fe_bw_mnitor;
+	uint32_t num_slc_be_bw_mnitor;
+} __packed;
+
+struct slc_sct_client_info {
+	uint32_t num_clients;
+	struct slc_mon_details slc_mon_info;
+	struct slc_client_details client;
+} __packed;
+
+
 struct qcom_slc_mon_data {
 	struct slc_partid_info part_info;
 	struct slc_capacity cap_stats;
 	struct slc_read_miss_cntr rd_miss_stats;
 } __packed;
 
-struct qcom_slc_mon_mem {
+struct slc_mon_data_values {
+	uint16_t cntr_index;
+	uint16_t mon_enabled;
+	uint32_t num_cache_lines;
+	uint64_t rd_misses;
+	uint64_t fe_rd_bytes;
+	uint64_t fe_wr_bytes;
+	uint64_t be_rd_bytes;
+	uint64_t be_wr_bytes;
+	uint64_t rd_hits;
+	uint64_t rd_incr;
+	uint64_t wr_hits;
+	uint64_t wr_incr;
+} __packed;
+
+struct qcom_slc_mon_data_v1 {
+	struct slc_partid_info part_info;
+	uint16_t cntr_index;
+	uint16_t mon_enabled;
+	uint32_t num_cache_lines;
+	uint64_t rd_misses;
+	uint64_t fe_rd_bytes;
+	uint64_t fe_wr_bytes;
+	uint64_t be_rd_bytes;
+	uint64_t be_wr_bytes;
+	uint64_t rd_hits;
+	uint64_t rd_incr;
+	uint64_t wr_hits;
+	uint64_t wr_incr;
+} __packed;
+
+#define SLC_NUM_PARTIDS		5
+struct qcom_slc_mon_mem_v0 {
 	uint32_t match_seq;
 	uint16_t msc_id;
 	uint16_t num_active_mon;
@@ -118,51 +235,96 @@ struct qcom_slc_mon_mem {
 	uint64_t last_capture_time;
 } __packed;
 
-struct capacity_info  {
-	uint32_t num_cache_lines;
+struct qcom_slc_mon_mem_v1 {
+	uint32_t match_seq;
+	uint16_t msc_id;
+	uint16_t num_active_mon;
 	uint64_t last_capture_time;
+	uint64_t slc_mpam_monitor_size;
+	struct qcom_slc_mon_data data[];
 } __packed;
 
-struct miss_info  {
-	uint64_t num_rd_misses;
+/* Needs to be deprecated once FW switched to V3 */
+struct qcom_slc_mon_mem_v2 {
+	uint32_t match_seq;
+	uint16_t msc_id;
+	uint16_t num_active_mon;
 	uint64_t last_capture_time;
+	uint64_t slc_mpam_monitor_size;
+	struct qcom_slc_mon_data_v1 data[];
 } __packed;
 
-union mon_values {
-	struct capacity_info capacity;
-	struct miss_info misses;
+union qcom_slc_monitor_memory {
+	struct qcom_slc_mon_mem_v0 mem_v0;
+	struct qcom_slc_mon_mem_v1 mem_v1;
+	struct qcom_slc_mon_mem_v2 mem_v2;
 };
 
-struct qcom_slc_mon_data_val {
-	struct slc_partid_info part_info;
+/* slc Monitor capability */
+struct slc_mon_capability {
+	uint32_t read_miss_config_available;
+	uint32_t capacity_config_available;
+};
+
+struct slc_mon_configured {
+	uint32_t read_miss_configured;
+	uint32_t capacity_configured;
+};
+
+/* msc slc capability */
+struct slc_client_capability {
+	struct slc_client_info client_info;
+	union slc_partid_capability_def *slc_partid_cap;
+	uint8_t enabled;
+	char *client_name;
+} __packed;
+
+struct qcom_slc_capability {
+	uint32_t num_clients;
+	struct slc_client_capability *slc_client_cap;
+	struct slc_mon_capability slc_mon_list;
+	struct slc_mon_configured slc_mon_configured;
+	struct qcom_slc_firmware_version firmware_ver;
+	uint32_t num_partids;
+} __packed;
+
+/* slc mon API parameters */
+struct capacity_info  {
+	uint32_t slc_mon_function;
+	uint64_t last_capture_time;
 	uint32_t num_cache_lines;
-	uint64_t rd_misses;
 } __packed;
 
-struct qcom_msc_slc_mon_val {
-	struct qcom_slc_mon_data_val data[SLC_NUM_PARTIDS];
+struct miss_info  {
+	uint32_t slc_mon_function;
 	uint64_t last_capture_time;
+	uint64_t num_rd_misses;
 } __packed;
 
-enum slc_clients_id {
-	APPS,
-	GPU,
-	NSP,
-	SLC_CLIENT_MAX,
-};
+struct slc_fe_stats  {
+	uint32_t slc_mon_function;
+	uint64_t last_capture_time;
+	uint64_t slc_fe_bytes;
+} __packed;
 
-enum gear_val {
-	GEAR_HIGH,
-	GEAR_LOW,
-	GEAR_BYPASS,
-	GEAR_MAX,
-};
+struct slc_be_stats  {
+	uint32_t slc_mon_function;
+	uint64_t last_capture_time;
+	uint64_t slc_be_bytes;
+} __packed;
 
-static char gear_index[][25] = {
-	"SLC_GEAR_HIGH",
-	"SLC_GEAR_LOW",
-	"SLC_GEAR_BYPASS",
-	"",
-};
+struct mon_ref {
+	uint32_t slc_mon_function;
+	uint64_t last_capture_time;
+	uint64_t mon_data;
+} __packed;
+
+union mon_values {
+	struct capacity_info capacity;
+	struct miss_info misses;
+	struct slc_fe_stats fe_stats;
+	struct slc_be_stats be_stats;
+	struct mon_ref ref;
+} __packed;
 
 #endif /* _QCOM_MPAM_SLC_H */
-- 
2.43.0

